Ww/issue 253 message buffer improvements

withinwheels opened the pull request:
Message display and archive is reworked.  Identical messages sometimes "collapse", meaning they are dropped in favor of increasing a displayed repetition count up to 100.

More specifically, an uninterrupted run of a single repeated non-combat message (`You finish your detailed search of the area.`) will collapse (`You finish your detailed search of the area. (x3)`).

Separately, identical combat messages (`The jackel missed you.`) arriving on the same player turn will collapse, in addition to their semi-colon folding, regardless of their arrival order within that turn (`The rat missed you (x3); The monkey missed you; The kobold missed you.`).

To do this, the message handling logic has been gathered more centrally in IO.c.  To continue expressing the various nuances such as requiring acknowledgment or folding combat message with semi-colons, a new flags enum has been introduced named `messageFlags`.

The ideas of a displayed message and an archived message have also drifted apart.  Displayed messages are still color encoded strings.  Archived messages are now structs equipped with a repetition count, turn observed and a flags vector in addition to their message.  The size of the archive in structs has also been separated from the number of lines in the archive display, since now individual combat messages each receive an entry in the archive.  Prior they were entered after semi-colon folding.

Message archive formatting is now performed by `formatMessageArchive`, which fills a given line-oriented buffer of a given size from the the bottom (with most recent entries) upwards.  This is then called once in `displayMessageArchive` (the sliding animation works only with screen buffers.)  It is called once also in the new `displayRecentMessages`, which redraws the "normal" three line message display area, with the option of confirming (or not) the latest messages, determined by player turn.

Finally, a number of small independent bug fixes are included as separate commits near the front of the PR.  These can be broken out into separate PRs (and retargeted to release) if needed.  The other commits are meant to be consumed one at a time in order.  Each should compile to produce a working game, and successively build up the new features.

withinwheels commented on src/brogue/IO.c:
```diff
@@ -3168,19 +3218,42 @@ void messageWithoutCaps(char *msg, enum messageFlags flags) {
+        archiveEntry = getArchivedMessage(i);
+
+        // Always consider the latest entry for collapse, but also past ones that arrived this turn
+        if (i > 1 && archiveEntry->turn != rogue.playerTurnNumber) {
```
Commit message not withstanding, this iteration of the logic is still here because we don't yet have the flags to distinguish combat messages from non-combat ones.  That arrives in the final commit, which then amends this line.

pender commented:
Works great in my test run! Nice change. And apparently way more cases to be handled than I'd imagined!

withinwheels commented:
Thanks!  The meat came down to moving the various formatting bits "behind" the message archive rather than "in front" of it, so that the archive itself remains unformatted.

Yesterday I took a stab at adding scrolling functionality.  It looks like it will easily piggy-back off this change, by allowing us to `formatMessageArchive` a buffer _larger_ than `MESSAGE_ARCHIVE_LINES` and then choosing the desired viewport.  Which in turn rather looks like the same machinery that's already used for the sliding animation.  I'll probably leave that to another pull request in order to do it well (and this one is busy enough!)

tmewett commented on src/brogue/IO.c:
```diff
@@ -3168,19 +3339,44 @@ void messageWithoutCaps(char *msg, enum messageFlags flags) {
+    newMessage = true;
+
+    // For each at most MESSAGE_ARCHIVE_ENTRIES - 1 past entries..
+    for (i = 1; i < MESSAGE_ARCHIVE_ENTRIES; i++) {
```
I'm struggling to understanding what this loop body does, or the function in general

withinwheels replied:
This is a pre-existing function, whose purpose before and after is to insert new messages into the message archive.

On [`master`][0], we can inspect its anatomy and conclude:
* It manages the "confirmed" (is bolded or not) state of the normal three-line message display, and shifts up to make room for the new message, which it then inserts.
* It inserts the message into the archive, which is a circular buffer.
* Get acknowledgement (show `--MORE--` and wait) from the player if required

[0]: https://github.com/tmewett/BrogueCE/blob/master/src/brogue/IO.c#L3154-L3192

On this branch:
* It first walks the archive, looking for a suitable message to collapse the new one into.  If one is found, its count is bumped.  This is the loop you have called out.
* If no suitable message to collapse into was found, instead initialize and insert the incoming message into the archive (again, dealing with the circular buffer) as a new one.  Only at this point, when we know a new row will be shown in the three-line display and in the archive, do we shift up.
* Again get acknowledgement if required, though `requireAcknowledgment` has become `REQUIRE_ACKNOWLEDGMENT`.

`updateMessageDisplay` is also replaced by `displayRecentMessages(false)`.  The `false` in particular means that newly arrived messages will be unconfirmed (shown in bold.)  It occurs to me that we can probably drop the `messageConfirmed` shifting now.

tmewett commented on src/brogue/IO.c:
```diff
@@ -3168,19 +3339,44 @@ void messageWithoutCaps(char *msg, enum messageFlags flags) {
+        // Consider messages that arrived this turn for collapsing.  Also
+        // consider the latest entry (which may be from past turns) if the
+        // incoming message is not semi-colon foldable.
+        if ((i > 1 || flags & FOLDABLE) && archiveEntry->turn != rogue.playerTurnNumber) {
```
The comment and the if are written inverted to each other, it's not easy to parse

withinwheels replied:
I will invert the conditional with an explicit negation.

tmewett reviewed (commented):
I'm sorry if I'm being dense but I'm having a hard time reasoning about this. There are many concepts; there is the message archive, there's the output buffer, there's lines, recent messages, confirmation, folding, splitting, ... Functions which I feel could do with more explanation / summary docs are:

* displayCombatText
* combatMessage
* formatRecentMessages (output args)
* foldMessages (output arg)
* messageWithoutCaps

tmewett commented on src/brogue/Rogue.h:
```diff
@@ -2608,6 +2610,19 @@ typedef struct buttonState {
 
+enum messageFlags {
+    REQUIRE_ACKNOWLEDGMENT        = Fl(0),
+    REFRESH_SIDEBAR               = Fl(1),
```
Do you know the need for this flag? Is it bad if we always refresh when a message is drawn?

withinwheels replied:
It ties into the next commit, 58b1fc6 , where we replace a use of `message` with `temporaryMessage` in `getInputTextString`.  `temporaryMessage` declares itself meant for prompts and, indeed, using `message` in `getInputTextString` to draw the prompt was causing problems.  

`getInputTextString` is used to prompt for various filenames, including in screens where the main game interface is not displayed.  As such we would not want it to (always) redraw the sidebar.  Notice that 

`temporaryMessage` used to redraw the sidebar unconditionally, which may have been the main reason it was not used in `getInputTextString`.  The sidebar-redrawing behavior of `temporaryMessage` is important, I suspect, in its uses for prompting the player to choose a direction for throwing or zapping a staff.

I suspect the work of threading yet another positional boolean argument (a la `requireAcknowledgement`) may have dissuaded earlier development going back and cleaning up what otherwise seems to me to be a pretty clear use for `temporaryMessage`.  However with `messageFlags` that heavy lifting is somewhat easier.

So, `REFRESH_SIDEBAR` now tells `temporaryMessage` whether or not to redraw the sidebar, enabling it for use in prompts where we don't want it seen.  Since `mesasge` also can conditionally redraw the sidebar, it pays attention to the flag as well.

withinwheels commented:
> Functions which I feel could do with more explanation / summary docs are: ...
Sure, I can put some more explanatory bread-crumbs in.

withinwheels commented on src/brogue/IO.c:
```diff
@@ -3050,24 +3050,90 @@ int formatCountedMessage(char *buffer, size_t size, archivedMessage *m) {
 // Display recent archived messages
-void displayRecentMessages() {
-    short i;
+void displayRecentMessages(boolean confirmLatest) {
```
It's a pre-existing concept which I've moved around a bit.  Notice that in the normal three-line message area (when the `M` display is closed) that recently arrived messages are bold and bright, and a little later they are dimmed.  Such bright messages are "unconfirmed".

This `displayRecentMessages` function is now responsible for redrawing the three-line message area, and so wants to know if the "most recent" (whatever that means) messages should be bright ("are unconfirmed") or not.  I opted to expose it as a parameter.

tmewett replied:
This feels like another function, some "confirmMessages", but i see the difficulty, because you can't know latest messages before you draw them

withinwheels replied:
..you know, there is a `confirmMessages`, which amounts to a call to `displayRecentMessages(true)`.

The original issue which led to `displayRecentMessages` was an inability to express `displayRecentMessages(false)` (`confirmMessages` can't do that.)

I may go and either remove `confirmMessages` or make it a delegate to `displayRecentMessages`.

tmewett replied:
It almost feels like confirmation should be a part of the archedMessage type. Instead of being stored by line in a different global. This would untangle the interplay between displaying and confirming, also removing the need for latestMessageLines I think?

withinwheels replied:
The confirmation flags could certainly be moved into the archive itself (I suppose I would put it in `flags` as either a new `CONFIRMED` or perhaps `UNCONFIRMED` value.)

Currently this branch redefines "confirmed" to be `archivedMessage.turn < rogue.playerTurnNumber` instead of letting the `messageConfirmed` global be the source of truth.  So in some sense a `UNCONFIRMED` flag is redundant with `.turn` itself.

The `confirmMessages` function (which has a few sprinkled call sites) would then end up walking the archive, ensuring the flags are correct on each message back into history.  For a state that's only supposed to exist for a brief while before being clear essentially forever, that walking might feel a bit schlemiel-ish.

What we might do is just retire the `messageConfirmed` global completely.  I believe the branch is pretty close to being able to do that.  I'm all for taking globals out behind the chemical shed and "retiring" them, I just didn't want to bake in even more complexity off the bat.

withinwheels replied:
..I just realized my train of thought had an important flaw.  "Confirmed" is still a single bit that twiddles back and forth every turn as the player takes actions within a turn that dims the message display.

We couldn't retire `messageConfirmed`, but we could change it from an array back to a single flag.

withinwheels replied:
..ok.  And if `messageConfirmed` was a single global boolean, `displayRecentMessages` might reference it directly instead of accepting a parameter.

Then `confirmMessages`, which would continue to exist, would set the global flag to the appropriate state and then delegate to `displayRecentMessages`.

tmewett replied:
> Currently this branch redefines "confirmed" to be archivedMessage.turn < rogue.playerTurnNumber instead of letting the messageConfirmed global be the source of truth

I hadn't fully realised this - in that case the argument to dRM makes sense: it's whether to ignore the lack of confirmation or not for just the drawn turn. In that case, yes I think it would be great to try and get this down to one concept, instead of the control flow (and where confirmLatest is true or false) having an implicit dependence on the turn progression in the game.

withinwheels replied:
Ok, I'll take another pass at cleaning it up.  I think I'll try the design I alluded to above (the array becomes a single boolean, cM still exists and defers to dRM, dRM loses its argument and just looks at the global boolean.)

withinwheels replied:
Well, instead of a boolean it became a short (`messagesUnconfirmed`), but otherwise that played out more or less as envisioned.

withinwheels commented on src/brogue/Combat.c:
```diff
@@ -1265,28 +1266,46 @@ void combatMessage(char *theMsg, color *theColor) {
+    length = strlen(combatText);
+
+    // Buffer combat messages here just for timing
+    if (length + strlen(newMsg) > COLS * 2 - 2) {
```
I will flush out the comments a bit then.  There are two aspects, and I'm not sure which you mean.

Firstly, this is a bounds check.  The `- 2` accounts for the newline that `snprintf` will include a few lines down, as well as the null terminator that `strlen` will not count.

Secondly, why is the buffering there at all?  I experimented with removing it, and found that it reversed the order of arrival of combat messages.  In a combat round the player is accustomed to seeing their own action first, followed by the actions of monsters and other things.  For whatever reason, removing the logical delay implied by this buffer write (and eventual flush) seems to reverse that, which I thought would be jarring.  Rather than untangle the logic I opted to leave it alone.

tmewett replied:
Thanks, I understand the buffering, I just think the comment is backwards, i.e. the comment says we are buffering messages but it's above a block which is emptying the buffer in case it over flows? It just seems a confusing place to annotate this. But it's not a problem, the function comment sets the stage for buffering now so it's clearer now :+1:

withinwheels commented on src/brogue/IO.c:
```diff
@@ -3108,7 +3108,9 @@ void temporaryMessage(char *msg, enum messageFlags flags) {
 
-    refreshSideBar(-1, -1, false);
+    if (flags & REFRESH_SIDEBAR) {
+      refreshSideBar(-1, -1, false);
```
nit: indent

withinwheels commented:
Rebased off latest master, and more comments have been added.

I also dropped the `messageConfirmed` shifting that I hadn't noticed was now extraneous, and inverted the condition in `messageWithoutCaps` to more closely match the prose.

tmewett commented on src/brogue/IO.c:
```diff
@@ -3153,34 +3329,62 @@ void flavorMessage(char *msg) {
+        // Consider messages that arrived this turn for collapsing.  Also
+        // consider the latest entry (which may be from past turns) if the
+        // incoming message is not semi-colon foldable.
+        if (!((i == 1 && !(flags & FOLDABLE)) || archiveEntry->turn == rogue.playerTurnNumber)) {
```
Why can't foldable messages collapse with the latest one?

withinwheels replied:
There's not a hard technical reason (though there is a soft one.)  The soft technical reason is that currently some function interfaces assume all folded messages occurred on the same player turn.  I could refactor around needing that assumption.

The real reason is that combat messages are the only foldable messages right now, and collapsing combat messages across turns struck me as perhaps too confusing to the player.  By breaking them up along turn lines we still get a clean delineation of rounds.

It might be better to extract a `COMBAT` value `messageType`, and set it to `FOLDABLE`, then explicitly talk about not collapsing `COMBAT` messages across turn boundaries.  Early versions of the branch did that, but it felt (slightly) like overkill and I dropped it.

tmewett replied:
Oh I see, that makes sense, it's good as it is :+1:

tmewett commented:
Thanks for your patience and changes, I wasn't familiar with the old code so it's a bit tough! I'm overall happy with it, just had one or two Qs left above

tmewett commented:
Thanks for that! Code looks good now.

One thing I noticed from testing right now is that folded message have capitals in each part - `You hit the kobold; The kobold hits you.` Would this be hard to fix?

withinwheels commented:
Ok, capitalization fixes are in and have all been absorbed into 2d1ea66, the last meaty commit.

The short story is the capitalization (& the quoting logic) has been extracted to `capitalizeAndPunctuateSentences`, and `message` ate `messageWithoutCaps` (since `messageWithoutCaps` now officially has nothing to do with capitalization.)

Like the other formatting passes, `capitalizeAndPunctuateSentences` is called by `formatRecentMessages` for messages coming out the archive, rather than by `message` for ones going in.

tmewett commented:
I think this is in a great state now. Thank you!

withinwheels commented:
Thank you for your diligent reviews!  I appreciate the time those take.  No rest for the wicked however: #295 is hot on the heels.
